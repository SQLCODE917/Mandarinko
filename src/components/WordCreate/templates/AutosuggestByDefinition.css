.suggestionSpelling {
  color: white;
  font-size: 1.5rem;
}

.suggestionDefinitionList {
  list-style: none;
  padding-left: 1rem;
}

.suggestionDefinition {
  padding-bottom: .25rem; 
}

.suggestionDefinition:last-child {
  padding-bottom: 0;
}
