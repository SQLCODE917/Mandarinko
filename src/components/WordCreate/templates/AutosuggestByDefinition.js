import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Field } from 'redux-form'
import renderAutosuggestField from './renderAutosuggestField.js'
import {
  getSuggestions,
  renderSuggestion,
  getSuggestionValue
} from './ByDefinition.js'

export class AutosuggestByDefinition extends Component {
  render() {
    const { words } = this.props
    const fieldProps = {
      name: 'id',
      type: 'text',
      placeholder: 'Search by definition',
      component: renderAutosuggestField,
      getSuggestions: getSuggestions(words),
      renderSuggestion: renderSuggestion,
      getSuggestionValue: getSuggestionValue
    }
    return (
      <Field {...fieldProps} />
    )
  }
}

AutosuggestByDefinition.propTypes = {
  words: PropTypes.object
}

/* istanbul ignore next */
const mapStateToProps = ({ words }) => ( { words: words.words } )

export default connect(mapStateToProps)(AutosuggestByDefinition)
