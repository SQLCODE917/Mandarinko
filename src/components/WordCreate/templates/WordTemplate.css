.word {

}

.siblings {

}

.main {

}

.definition {

}

.derivation {

}

.children {

}
