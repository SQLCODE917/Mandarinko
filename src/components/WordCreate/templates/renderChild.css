@import './WordCreateCommon.css';

.childrenContainer {
  composes: fieldArray;
}

.childForm {
  composes: fieldArrayField;
}
