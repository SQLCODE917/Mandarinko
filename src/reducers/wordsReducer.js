import * as types from '../actions/actionTypes'

export const INITIAL_STATE = {
  top2kWordIds: [],
  words: {},
  loading: false,
  spacedRepetition: {
    currentWordId: void(0),
    again: [],
    good: [],
    easy: []
  },
  error: void(0)
}

export default function wordsReducer(state=INITIAL_STATE, action) {
  switch (action.type) {
    case types.ADD_WORD:
      const {id, word} = action
      return {
        ...state,
        words: {
          ...state.words,
          [id]: word
        }
      };
    case types.ADD_WORDS:
      const { words } = action;
      return {
        ...state,
        allWords: words
      }
    case types.SET_CURRENT_WORD_ID:
      return {
        ...state,
        spacedRepetition: {
          ...state.spacedRepetition,
          currentWordId: action.id
        }
      }
    case types.LOADING:
      return {
        ...state,
        loading: action.status
      }
    case types.SET_TOP_2K:
      return {
        ...state,
        top2kWordIds: action.wordIds
      }
      return action.wordIds
    case types.ERROR: {
      return {
        ...state,
        error: action.error
      }
    }
    default:
      return state
  }
}
