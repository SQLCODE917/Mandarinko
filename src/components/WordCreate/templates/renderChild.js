import { FormSection } from 'redux-form'
import * as c from '../../../constants'
import WordTemplate from './WordTemplate'
import styles from './renderChild.css'

const renderChild = ({ fields, meta: { error, submitFailed } }) => (
  <section className={styles.childrenContainer}>
    <button type="button"
      title="Add Child"
      onClick={() => fields.push({})}>
      {c.plusSign} Child
    </button>
    {submitFailed && error && <span>{error}</span>}
    {fields.map((wordObject, index) => (
      <article className={styles.childForm}
        key={index}>

        <button type="button"
          title="Remove Child"
          onClick={() => fields.remove(index)}
        >{c.minusSign}</button>
        <FormSection name={`${wordObject}`}>
          <WordTemplate />
        </FormSection>
      </article>
    ))}
  </section>
);

export default renderChild 
