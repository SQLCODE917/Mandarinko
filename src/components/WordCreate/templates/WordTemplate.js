import { Field, FieldArray } from 'redux-form'
import renderSibling from './renderSibling'
import renderSpelling from './renderSpelling'
import renderInputField from './renderInputField'
import renderDefinition from './renderDefinition'
import renderChild from './renderChild'
import styles from './WordTemplate.css'

const WordTemplate = () => (
  <article className={styles.word}>
    <section className={styles.siblings}>
      <FieldArray
        name="siblings"
        component={renderSibling}
      />
    </section>
    <article className={styles.main}>
      <FieldArray
        name="spelling"
        component={renderSpelling}
      />
      <Field
        name="pronounciation"
        component={renderInputField}
        type="text"
        label="Pronounciation"
      />
      <section className={styles.definition}>
        <FieldArray
          name="definition"
          component={renderDefinition}
        />
      </section>
    </article>
    <section className={styles.derivation}>
      <Field
        name="derivation"
        component={renderInputField}
        type="text"
        label="Derivation"
      />
    </section>
    <section className={styles.children}>
      <FieldArray
        name="children"
        component={renderChild}
      />
    </section>
  </article>
)

export default WordTemplate
