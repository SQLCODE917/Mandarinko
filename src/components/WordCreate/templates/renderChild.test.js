import React from 'react'
import { mount } from 'enzyme'
import {
  FieldArray,
  Form,
  reducer as formReducer,
  reduxForm
} from 'redux-form'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import {
  combineReducers,
  createStore,
  applyMiddleware
} from 'redux'

import wordsReducer from '../../../reducers/wordsReducer'
import renderChild from './renderChild'
import AutosuggestByDefinition from './AutosuggestByDefinition'
import WordTemplate from './WordTemplate'
import ErrorLine from './errorLine'

describe('renderChild', () => {
  let store, submitSpy, wrapper

  beforeEach(() => {
    submitSpy = jest.fn()
    store = createStore(
      combineReducers({
        form: formReducer,
        words: wordsReducer
      }),
      { form: {}, words: { words: {} } },
      applyMiddleware(thunk)
    )
    const form = ({ onSubmit }) => 
      <Form onSubmit={onSubmit}>
        <FieldArray
          name="children"
          component={renderChild}
        />
      </Form>
    const WrappedForm = reduxForm({
      form: 'testForm',
      onSubmit: submitSpy 
    })(form)
    wrapper = mount(
      <Provider store={store}>
        <WrappedForm />
      </Provider>
    )

  })

  it('renders a blank word form', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('adds a lookup child form when clicked', () => {
    wrapper.find('[title="Look Up a Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(1)
    expect(wrapper.find(AutosuggestByDefinition)).toHaveLength(1)
  })

  it('adds a child form when clicked', () => {
    wrapper.find('[title="Add Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(1)
    expect(wrapper.find(WordTemplate)).toHaveLength(1)
  })

  it('removes the added child form when clicked', () => {
    wrapper.find('[title="Add Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(1)
    wrapper.find('[title="Remove Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(0)
  })

  it('removes the lookup child form when clicked', () => {
    wrapper.find('[title="Look Up a Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(1)
    expect(wrapper.find(AutosuggestByDefinition)).toHaveLength(1)
    wrapper.find('[title="Remove Child"]').simulate('click')
    wrapper.update()
    expect(wrapper.find('.childForm')).toHaveLength(0)
  })

  it('shows the error when necessary', () => {
    const props = {
      fields: [],
      meta: {
        error: 'ayyyy',
        submitFailed: true
      }
    }
    const wrapper = mount(<div>{renderChild(props)}</div>)
    expect(wrapper.find(ErrorLine)).toHaveLength(1)
  })
})
