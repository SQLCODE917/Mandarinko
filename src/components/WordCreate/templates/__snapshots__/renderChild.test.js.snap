// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderChild renders a blank word form 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <ReduxForm>
    <Connect(Form(form))
      destroyOnUnmount={true}
      enableReinitialize={false}
      forceUnregisterOnUnmount={false}
      form="testForm"
      getFormState={[Function]}
      keepDirtyOnReinitialize={false}
      onSubmit={[MockFunction]}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldError={[Function]}
      shouldValidate={[Function]}
      shouldWarn={[Function]}
      touchOnBlur={true}
      touchOnChange={false}
      updateUnregisteredFields={false}
    >
      <Form(form)
        anyTouched={false}
        array={
          Object {
            "insert": [Function],
            "move": [Function],
            "pop": [Function],
            "push": [Function],
            "remove": [Function],
            "removeAll": [Function],
            "shift": [Function],
            "splice": [Function],
            "swap": [Function],
            "unshift": [Function],
          }
        }
        arrayInsert={[Function]}
        arrayMove={[Function]}
        arrayPop={[Function]}
        arrayPush={[Function]}
        arrayRemove={[Function]}
        arrayRemoveAll={[Function]}
        arrayShift={[Function]}
        arraySplice={[Function]}
        arraySwap={[Function]}
        arrayUnshift={[Function]}
        asyncValidating={false}
        autofill={[Function]}
        blur={[Function]}
        change={[Function]}
        clearAsyncError={[Function]}
        clearFields={[Function]}
        clearSubmit={[Function]}
        clearSubmitErrors={[Function]}
        destroy={[Function]}
        destroyOnUnmount={true}
        dirty={false}
        dispatch={[Function]}
        enableReinitialize={false}
        focus={[Function]}
        forceUnregisterOnUnmount={false}
        form="testForm"
        getFormState={[Function]}
        initialize={[Function]}
        initialized={false}
        invalid={false}
        keepDirtyOnReinitialize={false}
        onSubmit={[MockFunction]}
        persistentSubmitErrors={false}
        pristine={true}
        pure={true}
        registerField={[Function]}
        registeredFields={
          Object {
            "children": Object {
              "count": 1,
              "name": "children",
              "type": "FieldArray",
            },
          }
        }
        reset={[Function]}
        resetSection={[Function]}
        setSubmitFailed={[Function]}
        setSubmitSucceeded={[Function]}
        shouldAsyncValidate={[Function]}
        shouldError={[Function]}
        shouldValidate={[Function]}
        shouldWarn={[Function]}
        startAsyncValidation={[Function]}
        startSubmit={[Function]}
        stopAsyncValidation={[Function]}
        stopSubmit={[Function]}
        submit={[Function]}
        submitFailed={false}
        submitSucceeded={false}
        submitting={false}
        syncErrors={Object {}}
        syncWarnings={Object {}}
        touch={[Function]}
        touchOnBlur={true}
        touchOnChange={false}
        unregisterField={[Function]}
        untouch={[Function]}
        updateSyncErrors={[Function]}
        updateSyncWarnings={[Function]}
        updateUnregisteredFields={false}
        valid={true}
        validExceptSubmit={true}
        values={Object {}}
      >
        <form
          anyTouched={false}
          array={
            Object {
              "insert": [Function],
              "move": [Function],
              "pop": [Function],
              "push": [Function],
              "remove": [Function],
              "removeAll": [Function],
              "shift": [Function],
              "splice": [Function],
              "swap": [Function],
              "unshift": [Function],
            }
          }
          asyncValidate={[Function]}
          asyncValidating={false}
          autofill={[Function]}
          blur={[Function]}
          change={[Function]}
          clearAsyncError={[Function]}
          clearFields={[Function]}
          clearSubmit={[Function]}
          clearSubmitErrors={[Function]}
          destroy={[Function]}
          dirty={false}
          dispatch={[Function]}
          form="testForm"
          handleSubmit={[Function]}
          initialize={[Function]}
          initialized={false}
          invalid={false}
          onSubmit={[MockFunction]}
          pristine={true}
          pure={true}
          reset={[Function]}
          resetSection={[Function]}
          submit={[Function]}
          submitFailed={false}
          submitSucceeded={false}
          submitting={false}
          touch={[Function]}
          untouch={[Function]}
          valid={true}
        >
          <Form
            onSubmit={[MockFunction]}
          >
            <form
              onSubmit={[MockFunction]}
            >
              <FieldArray
                component={[Function]}
                name="children"
              >
                <Connect(ConnectedFieldArray)
                  _reduxForm={
                    Object {
                      "anyTouched": false,
                      "array": Object {
                        "insert": [Function],
                        "move": [Function],
                        "pop": [Function],
                        "push": [Function],
                        "remove": [Function],
                        "removeAll": [Function],
                        "shift": [Function],
                        "splice": [Function],
                        "swap": [Function],
                        "unshift": [Function],
                      },
                      "arrayInsert": [Function],
                      "arrayMove": [Function],
                      "arrayPop": [Function],
                      "arrayPush": [Function],
                      "arrayRemove": [Function],
                      "arrayRemoveAll": [Function],
                      "arrayShift": [Function],
                      "arraySplice": [Function],
                      "arraySwap": [Function],
                      "arrayUnshift": [Function],
                      "asyncErrors": undefined,
                      "asyncValidate": [Function],
                      "asyncValidating": false,
                      "autofill": [Function],
                      "blur": [Function],
                      "change": [Function],
                      "clearAsyncError": [Function],
                      "clearFields": [Function],
                      "clearSubmit": [Function],
                      "clearSubmitErrors": [Function],
                      "destroy": [Function],
                      "destroyOnUnmount": true,
                      "dirty": false,
                      "dispatch": [Function],
                      "enableReinitialize": false,
                      "error": undefined,
                      "focus": [Function],
                      "forceUnregisterOnUnmount": false,
                      "form": "testForm",
                      "getFormState": [Function],
                      "getValues": [Function],
                      "initialValues": undefined,
                      "initialize": [Function],
                      "initialized": false,
                      "invalid": false,
                      "keepDirtyOnReinitialize": false,
                      "onSubmit": [MockFunction],
                      "persistentSubmitErrors": false,
                      "pristine": true,
                      "pure": true,
                      "register": [Function],
                      "registerField": [Function],
                      "registerInnerOnSubmit": [Function],
                      "registeredFields": undefined,
                      "reset": [Function],
                      "resetSection": [Function],
                      "sectionPrefix": undefined,
                      "setSubmitFailed": [Function],
                      "setSubmitSucceeded": [Function],
                      "shouldAsyncValidate": [Function],
                      "shouldError": [Function],
                      "shouldValidate": [Function],
                      "shouldWarn": [Function],
                      "startAsyncValidation": [Function],
                      "startSubmit": [Function],
                      "stopAsyncValidation": [Function],
                      "stopSubmit": [Function],
                      "submit": [Function],
                      "submitFailed": false,
                      "submitSucceeded": false,
                      "submitting": false,
                      "syncErrors": Object {},
                      "syncWarnings": Object {},
                      "touch": [Function],
                      "touchOnBlur": true,
                      "touchOnChange": false,
                      "triggerSubmit": undefined,
                      "unregister": [Function],
                      "unregisterField": [Function],
                      "untouch": [Function],
                      "updateSyncErrors": [Function],
                      "updateSyncWarnings": [Function],
                      "updateUnregisteredFields": false,
                      "valid": true,
                      "validExceptSubmit": true,
                      "values": Object {},
                      "warning": undefined,
                    }
                  }
                  component={[Function]}
                  name="children"
                >
                  <ConnectedFieldArray
                    _reduxForm={
                      Object {
                        "anyTouched": false,
                        "array": Object {
                          "insert": [Function],
                          "move": [Function],
                          "pop": [Function],
                          "push": [Function],
                          "remove": [Function],
                          "removeAll": [Function],
                          "shift": [Function],
                          "splice": [Function],
                          "swap": [Function],
                          "unshift": [Function],
                        },
                        "arrayInsert": [Function],
                        "arrayMove": [Function],
                        "arrayPop": [Function],
                        "arrayPush": [Function],
                        "arrayRemove": [Function],
                        "arrayRemoveAll": [Function],
                        "arrayShift": [Function],
                        "arraySplice": [Function],
                        "arraySwap": [Function],
                        "arrayUnshift": [Function],
                        "asyncErrors": undefined,
                        "asyncValidate": [Function],
                        "asyncValidating": false,
                        "autofill": [Function],
                        "blur": [Function],
                        "change": [Function],
                        "clearAsyncError": [Function],
                        "clearFields": [Function],
                        "clearSubmit": [Function],
                        "clearSubmitErrors": [Function],
                        "destroy": [Function],
                        "destroyOnUnmount": true,
                        "dirty": false,
                        "dispatch": [Function],
                        "enableReinitialize": false,
                        "error": undefined,
                        "focus": [Function],
                        "forceUnregisterOnUnmount": false,
                        "form": "testForm",
                        "getFormState": [Function],
                        "getValues": [Function],
                        "initialValues": undefined,
                        "initialize": [Function],
                        "initialized": false,
                        "invalid": false,
                        "keepDirtyOnReinitialize": false,
                        "onSubmit": [MockFunction],
                        "persistentSubmitErrors": false,
                        "pristine": true,
                        "pure": true,
                        "register": [Function],
                        "registerField": [Function],
                        "registerInnerOnSubmit": [Function],
                        "registeredFields": undefined,
                        "reset": [Function],
                        "resetSection": [Function],
                        "sectionPrefix": undefined,
                        "setSubmitFailed": [Function],
                        "setSubmitSucceeded": [Function],
                        "shouldAsyncValidate": [Function],
                        "shouldError": [Function],
                        "shouldValidate": [Function],
                        "shouldWarn": [Function],
                        "startAsyncValidation": [Function],
                        "startSubmit": [Function],
                        "stopAsyncValidation": [Function],
                        "stopSubmit": [Function],
                        "submit": [Function],
                        "submitFailed": false,
                        "submitSucceeded": false,
                        "submitting": false,
                        "syncErrors": Object {},
                        "syncWarnings": Object {},
                        "touch": [Function],
                        "touchOnBlur": true,
                        "touchOnChange": false,
                        "triggerSubmit": undefined,
                        "unregister": [Function],
                        "unregisterField": [Function],
                        "untouch": [Function],
                        "updateSyncErrors": [Function],
                        "updateSyncWarnings": [Function],
                        "updateUnregisteredFields": false,
                        "valid": true,
                        "validExceptSubmit": true,
                        "values": Object {},
                        "warning": undefined,
                      }
                    }
                    arrayInsert={[Function]}
                    arrayMove={[Function]}
                    arrayPop={[Function]}
                    arrayPush={[Function]}
                    arrayRemove={[Function]}
                    arrayRemoveAll={[Function]}
                    arrayShift={[Function]}
                    arraySplice={[Function]}
                    arraySwap={[Function]}
                    arrayUnshift={[Function]}
                    component={[Function]}
                    dirty={false}
                    length={0}
                    name="children"
                    pristine={true}
                    rerenderOnEveryChange={false}
                  >
                    <renderChild
                      fields={
                        Object {
                          "_isFieldArray": true,
                          "forEach": [Function],
                          "get": [Function],
                          "getAll": [Function],
                          "insert": [Function],
                          "length": 0,
                          "map": [Function],
                          "move": [Function],
                          "name": "children",
                          "pop": [Function],
                          "push": [Function],
                          "reduce": [Function],
                          "remove": [Function],
                          "removeAll": [Function],
                          "shift": [Function],
                          "swap": [Function],
                          "unshift": [Function],
                        }
                      }
                      meta={
                        Object {
                          "dirty": false,
                          "error": undefined,
                          "form": "testForm",
                          "invalid": false,
                          "pristine": true,
                          "submitFailed": undefined,
                          "submitting": undefined,
                          "valid": true,
                          "warning": undefined,
                        }
                      }
                    >
                      <section
                        className="childrenContainer"
                      >
                        <button
                          className="button"
                          onClick={[Function]}
                          title="Add Child"
                          type="button"
                        >
                          +
                           Child
                        </button>
                        <button
                          className="button"
                          onClick={[Function]}
                          title="Look Up a Child"
                          type="button"
                        >
                          🔎
                           Child
                        </button>
                      </section>
                    </renderChild>
                  </ConnectedFieldArray>
                </Connect(ConnectedFieldArray)>
              </FieldArray>
            </form>
          </Form>
        </form>
      </Form(form)>
    </Connect(Form(form))>
  </ReduxForm>
</Provider>
`;
